lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+        }
+    }
+    if (defined($contigobj)) {
+        my $count = 0;
+        my $size = 0;
+        if (defined($contigobj->{contigs})) {
+            $inputgenome->{contigs} = $contigobj->{contigs};
+            for (my $i=0; $i < @{$inputgenome->{contigs}}; $i++) {
+                $count++;
+                $size += length($inputgenome->{contigs}->[$i]->{sequence});
+                $inputgenome->{contigs}->[$i]->{dna} = $inputgenome->{contigs}->[$i]->{sequence};
+                delete $inputgenome->{contigs}->[$i]->{sequence};
+            }
+        }
+    
+        if ($contigobj->{_kbasetype} eq "ContigSet") {
+            $inputgenome->{contigset_ref} = $contigobj->{_reference};
+        } else {
+            $inputgenome->{assembly_ref} = $contigobj->{_reference};
+        }
+
+        if ($is_assembly) {
+            $message .= "The RAST algorithm was applied to annotating a genome sequence comprised of ".$count." contigs containing ".$size." nucleotides. \nNo initial gene calls were provided.\n";
+        } else {
+            $message .= "The RAST algorithm was applied to annotating an existing genome: ".$parameters->{scientific_name}.". \nThe sequence for this genome is comprised of ".$count." contigs containing ".$size." nucleotides. \nThe input genome has ".@{$inputgenome->{features}}." existing coding features.\n and ".@{$inputgenome->{non_coding_features}}." existing non-coding features.\n";
+            $message .= "NOTE: Older input genomes did not properly separate coding and non-coding features.\n" if (@{$inputgenome->{non_coding_features}} == 0);
+        }       
+    } else {
+        if($inputgenome->{domain} !~ /Eukaryota|Plant/){
+            $message .= "The RAST algorithm was applied to annotating an existing genome: ".$parameters->{scientific_name}.". \nNo DNA sequence was provided for this genome, therefore new genes cannot be called. \nWe can only functionally annotate the ".@{$inputgenome->{features}}." existing features.\n";
+        } else {
+            $message .= "The RAST algorithm was applied to functionally annotate ".@{$inputgenome->{features}}." coding features.\n and ".@{$inputgenome->{non_coding_features}}." existing non-coding features in an existing genome: ".$parameters->{scientific_name}.".\n";
+            $message .= "NOTE: Older input genomes did not properly separate coding and non-coding features.\n" if (@{$inputgenome->{non_coding_features}} == 0);
+        }
+    }
+    if (defined($rast_details{types})) {
+        my %types = %{$rast_details{types}};
+        $message .= "Input genome has the following feature types:\n";
+        for my $key (sort keys(%types)) {
```

Comment:
```perl
for my $key ( sort keys %{ $rast_details{ types } } ) {
```
no need for the extra hash

Reply:
yes it's needed for the line inside the loop, and for readability